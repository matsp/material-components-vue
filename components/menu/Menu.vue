<template>
  <div
    class="mdc-menu mdc-menu-surface"
    @MDCMenu:selected="onSelect"
    @_init="onParentInit"
  >
    <slot />
  </div>
</template>

<script>
import { Corner, MDCMenu } from '@material/menu'
import { DefaultFocusState } from '@material/menu/constants'

import { baseComponentMixin, themeClassMixin } from '../base'
import { MDCComponent } from '@material/base/component'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  inject: {
    getMenu: {
      default: null
    }
  },
  provide () {
    return {
      getList: this.getList
    }
  },
  model: {
    prop: 'open',
    event: 'change'
  },
  props: {
    open: {
      type: Boolean,
      default: false
    },
    quickOpen: {
      type: Boolean,
      default: false
    },
    anchorCorner: {
      type: [Number, String],
      default: 0
    },
    absolutePositionX: {
      type: Number,
      default: Number.POSITIVE_INFINITY
    },
    absolutePositionY: {
      type: Number,
      default: Number.POSITIVE_INFINITY
    },
    hoistToBody: {
      type: Boolean,
      default: false
    },
    fixed: {
      type: Boolean,
      default: false
    },
    wrapFocus: {
      type: Boolean,
      default: true
    },
    defaultFocusState: {
      type: [Number, String],
      default: 1
    },
    selectedIndex: {
      type: Number,
      default: -1
    },
    isHoisted: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      mdcMenu: undefined,
      slotObserver: undefined
    }
  },
  computed: {
    focusState () {
      if (!this.defaultFocusState) return null

      const upperCaseFocusState = String(this.defaultFocusState).toUpperCase()
      if (isNaN(this.defaultFocusState)) {
        return DefaultFocusState[upperCaseFocusState]
      }

      const numberFocusState = Number(this.defaultFocusState)
      if (!isNaN(this.defaultFocusState)) {
        return numberFocusState
      }

      return null
    },
    _anchorCorner () {
      if (!this.anchorCorner) return null

      const upperCaseFocusState = String(this.anchorCorner).toUpperCase()
      if (isNaN(this.anchorCorner)) {
        return Corner[upperCaseFocusState]
      }

      const numberFocusState = Number(this.anchorCorner)
      if (!isNaN(this.anchorCorner)) {
        return numberFocusState
      }

      return null
    }
  },
  watch: {
    open (val) {
      if (this.mdcMenu.open !== val) this.mdcMenu.open = val
    },
    quickOpen () {
      this.mdcMenu.quickOpen = this.quickOpen
    },
    hoistToBody () {
      if (this.hoistToBody) {
        this.mdcMenu.hoistMenuToBody()
      }
    },
    isHoisted () {
      this.mdcMenu.setIsHoisted(this.isHoisted)
    },
    fixed () {
      this.mdcMenu.setFixedPosition(this.fixed)
    },
    anchorCorner () {
      if (this._anchorCorner !== null) {
        this.mdcMenu.setAnchorCorner(this.anchorCorner)
      }
    },
    absolutePositionX () {
      if (this.absolutePositionX !== Number.POSITIVE_INFINITY) {
        this.mdcMenu.setAbsolutePosition(this.absolutePositionX, this.absolutePositionY)
      }
    },
    absolutePositionY () {
      if (this.absolutePositionY !== Number.POSITIVE_INFINITY) {
        this.mdcMenu.setAbsolutePosition(this.absolutePositionX, this.absolutePositionY)
      }
    },
    wrapFocus () {
      this.mdcMenu.wrapFocus = this.wrapFocus
    },
    defaultFocusState () {
      if (this.focusState !== null) {
        this.mdcMenu.setDefaultFocusState(this.focusState)
      }
    },
    'mdcMenu.open' (val) {
      this.$emit('change', val)
    },
    selectedIndex () {
      if (this.selectedIndex >= 0) this.mdcMenu.setSelectedIndex(this.selectedIndex)
    }
  },
  mounted () {
    this.updateSlot()
    this.slotObserver = new MutationObserver(() => this.updateSlot())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    if (!(this.getMenu instanceof Function)) { // can not be init by parent
      this.instantiate()
    }
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
    if (this.mdcMenu instanceof MDCMenu) {
      this.mdcMenu.destroy()
    }
  },
  methods: {
    instantiate () {
      this.mdcMenu = MDCMenu.attachTo(this.$el)
      this.propsSetting()
      this.$nextTick(() => { // wait for the DOM change
        // tell all the children that the parent is initialized
        if (this.mdcMenu.list_ instanceof MDCComponent) {
          this.mdcMenu.list_.emit('_init')
        }
      })
    },
    updateSlot () {
      if (this.$slots.default) {
        this.$slots.default.map(n => {
          if (n.elm instanceof HTMLElement) {
            n.elm.setAttribute('role', 'menu')
            n.elm.setAttribute('aria-hidden', 'true')
            n.elm.setAttribute('aria-orientation', 'vertical')
            n.elm.querySelectorAll('.mdc-list-item').forEach(i => {
              i.setAttribute('role', 'menuitem')
            })
          }
        })
      }
    },
    onSelect (event) {
      this.model = false
      this.$emit('selected', event.detail)
    },
    onParentInit () {
      const menu = this.getMenu()
      if (menu instanceof MDCMenu) {
        if (this.mdcMenu instanceof MDCMenu) this.mdcMenu.destroy()
        this.mdcMenu = menu
        this.propsSetting()
      }
    },
    propsSetting () {
      if (this.mdcMenu instanceof MDCMenu) {
        if (this.hoistToBody) {
          this.mdcMenu.hoistMenuToBody()
        }
        this.mdcMenu.setFixedPosition(this.fixed)
        if (this._anchorCorner) {
          this.mdcMenu.setAnchorCorner(this._anchorCorner)
        }
        if (this.absolutePositionX !== Number.POSITIVE_INFINITY && this.absolutePositionY !== Number.POSITIVE_INFINITY) {
          this.mdcMenu.setAbsolutePosition(this.absolutePositionX, this.absolutePositionY)
        }
        this.mdcMenu.wrapFocus = this.wrapFocus

        if (this.focusState !== null) {
          this.mdcMenu.setDefaultFocusState(this.focusState)
        }
        this.mdcMenu.setIsHoisted(this.isHoisted)

        if (this.selectedIndex >= 0) this.mdcMenu.setSelectedIndex(this.selectedIndex)
      }
    },
    getList () {
      return this.mdcMenu.list_
    }
  }
}
</script>
