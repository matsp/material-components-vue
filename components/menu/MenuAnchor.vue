<template>
  <div class="mdc-menu-surface--anchor">
    <slot />
  </div>
</template>
