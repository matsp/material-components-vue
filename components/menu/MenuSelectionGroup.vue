<template>
  <ul class="mdc-menu__selection-group">
    <slot />
  </ul>
</template>

<script>
export default {
  data () {
    return {
      slotObserver: undefined
    }
  },
  mounted () {
    this.$nextTick(() => {
      this.updateSlot()
      this.slotObserver = new MutationObserver(() => this.updateSlot())
      this.slotObserver.observe(this.$el, {
        childList: true,
        subtree: true
      })
    })
  },
  methods: {
    updateSlot () {
      if (this.$slots.default) {
        this.$slots.default.forEach(n => {
          if (n.elm instanceof Element && n.elm.classList.contains('mdc-list-item')) {
            n.elm.querySelector('.mdc-list-item__graphic').classList.add('mdc-menu__selection-group-icon')
          }
        })
      }
    }
  }
}
</script>

<style scoped>

</style>
