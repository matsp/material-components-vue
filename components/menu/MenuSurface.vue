<template>
  <div
    class="mdc-menu-surface"
    @MDCMenuSurface:closed="onClosed"
    @MDCMenuSurface:opened="onOpened"
  >
    <slot />
  </div>
</template>

<script>
import { Corner, MDCMenuSurface } from '@material/menu-surface'

export default {
  model: {
    prop: 'open',
    event: 'change'
  },
  props: {
    open: {
      type: Boolean,
      default: false
    },
    fixed: {
      type: Boolean,
      default: false
    },
    quickOpen: {
      type: Boolean,
      default: false
    },
    anchorCorner: {
      type: [Number, String],
      default: 0
    },
    absolutePositionX: {
      type: Number,
      default: null
    },
    absolutePositionY: {
      type: Number,
      default: null
    },
    hoistToBody: {
      type: Boolean,
      default: false
    },
    isHoisted: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      mdcMenuSurface: undefined
    }
  },
  computed: {
    model: {
      get () {
        return this.open
      },
      set (value) {
        this.$emit('change', value)
      }
    },
    _anchorCorner () {
      if (!this.anchorCorner) return null

      const upperCaseFocusState = String(this.anchorCorner).toUpperCase()
      if (isNaN(this.anchorCorner)) {
        return Corner[upperCaseFocusState]
      }

      const numberFocusState = Number(this.anchorCorner)
      if (!isNaN(this.anchorCorner)) {
        return numberFocusState
      }

      return null
    }
  },
  watch: {
    open () {
      this.open ? this.mdcMenuSurface.getDefaultFoundation().open() : this.mdcMenuSurface.getDefaultFoundation().close()
    },
    quickOpen () {
      this.mdcMenuSurface.quickOpen = this.quickOpen
    },
    hoistToBody () {
      if (this.hoistToBody) {
        this.mdcMenuSurface.hoistMenuToBody()
      }
    },
    isHoisted () {
      this.mdcMenuSurface.setIsHoisted(this.isHoisted)
    },
    fixed () {
      this.mdcMenuSurface.setFixedPosition(this.fixed)
    },
    anchorCorner () {
      if (this._anchorCorner !== null) {
        this.mdcMenuSurface.setAnchorCorner(this.anchorCorner)
      }
    },
    absolutePositionX () {
      if (this.absolutePositionX !== null) {
        this.mdcMenuSurface.setAbsolutePosition(this.absolutePositionX, this.absolutePositionY)
      }
    },
    absolutePositionY () {
      if (this.absolutePositionY !== null) {
        this.mdcMenuSurface.setAbsolutePosition(this.absolutePositionX, this.absolutePositionY)
      }
    }
  },
  mounted () {
    this.$nextTick(() => {
      this.mdcMenuSurface = MDCMenuSurface.attachTo(this.$el)
      this.mdcMenuSurface.setFixedPosition(this.fixed)
      if (this._anchorCorner) {
        this.mdcMenuSurface.setAnchorCorner(this._anchorCorner)
      }
      if (this.absolutePositionX !== null || this.absolutePositionY !== null) {
        this.mdcMenuSurface.setAbsolutePosition(this.absolutePositionX, this.absolutePositionY)
      }
      if (this.hoistToBody) {
        this.mdcMenuSurface.hoistMenuToBody()
      }
      this.mdcMenuSurface.setIsHoisted(this.isHoisted)

      this.open ? this.mdcMenuSurface.getDefaultFoundation().open() : this.mdcMenuSurface.getDefaultFoundation().close()
    })
  },
  beforeDestroy () {
    this.mdcMenuSurface.destroy()
  },
  methods: {
    onClosed () {
      this.model = false
      this.$emit('closed')
    },
    onOpened () {
      this.model = true
      this.$emit('opened')
    }
  }
}
</script>
