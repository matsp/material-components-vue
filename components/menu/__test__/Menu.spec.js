import { mount } from '@vue/test-utils'
import Menu from '../Menu.vue'
import MenuAnchor from '../MenuAnchor.vue'
import MenuSurface from '../MenuSurface'
import MenuSelectionGroup from '../MenuSelectionGroup'

describe('Menu', () => {
  it('should mount', () => {
    const wrapper = mount(Menu)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(Menu)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-menu')
    expect(wrapper.classes()).toContain('mdc-menu-surface')
  })

  it('should render as fixed', () => {
    const wrapper = mount(Menu, {
      propsData: {
        fixed: true
      }
    })
    expect(wrapper).toMatchSnapshot()
  })
})

describe('MenuSurface', () => {
  it('should mount', () => {
    const wrapper = mount(MenuSurface)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(MenuSurface)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-menu-surface')
  })

  it('should render as fixed', () => {
    const wrapper = mount(MenuSurface, {
      propsData: {
        fixed: true
      }
    })
    expect(wrapper).toMatchSnapshot()
  })
})

describe('MenuAnchor', () => {
  it('should mount', () => {
    const wrapper = mount(MenuAnchor)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(MenuAnchor)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-menu-surface--anchor')
  })
})

describe('MenuSelectionGroup', () => {
  it('should mount', () => {
    const wrapper = mount(MenuSelectionGroup)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(MenuSelectionGroup)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-menu__selection-group')
  })
})
