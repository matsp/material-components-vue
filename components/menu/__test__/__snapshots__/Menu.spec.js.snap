// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render as fixed 1`] = `<div class="mdc-menu mdc-menu-surface mdc-menu-surface--fixed"></div>`;

exports[`Menu should render with no prop 1`] = `<div class="mdc-menu mdc-menu-surface"></div>`;

exports[`MenuAnchor should render with no prop 1`] = `<div class="mdc-menu-surface--anchor"></div>`;

exports[`MenuSelectionGroup should render with no prop 1`] = `<ul class="mdc-menu__selection-group"></ul>`;

exports[`MenuSurface should render as fixed 1`] = `<div class="mdc-menu-surface"></div>`;

exports[`MenuSurface should render with no prop 1`] = `<div class="mdc-menu-surface"></div>`;
